### PR TITLE
fix(core): Proper retry count in KazooRetry

### PR DESCRIPTION
**Pull request by ceache**
Make sure the number of attempts matches the `max_retry` parameter. Add unit tests to that effect.

## Why is this needed?

So that users can precisely control the number of retries

## Proposed Changes

  - Move the location of the attempt increase to before the check against `max_try` so we don't end up calling `max_tries + 1` times the function.

## Does this PR introduce any breaking change?

No, it shouldn't.

### CODE REVIEW

File: kazoo/tests/test_retry.py [hunk 1]
```diff
@@ -1,89 +1,107 @@
-import unittest
-
 import pytest
 
 
-class TestRetrySleeper(unittest.TestCase):
```
Comment (StephenSorriaux) [inline kazoo/tests/test_retry.py:6]: Can I ask why not using a class anymore? It seems to make the `_try_counts` handling a little more "simple" (ie. without the use of `global` + doing its reinit in the `setup()` method) but I might be missing something.

File: kazoo/retry.py [hunk 2]
```diff
@@ -133,10 +133,10 @@ def __call__(self, func, *args, **kwargs):
             except ConnectionClosedError:
                 raise
             except self.retry_exceptions:
+                self._attempts += 1
                 # Note: max_tries == -1 means infinite tries.
                 if self._attempts == self.max_tries:
                     raise RetryFailedError("Too many retry attempts")
-                self._attempts += 1
```
Comment (StephenSorriaux) [inline kazoo/retry.py:139]: Great catch!

File: kazoo/retry.py [hunk 3]
```diff
@@ -133,10 +133,10 @@ def __call__(self, func, *args, **kwargs):
             except ConnectionClosedError:
                 raise
             except self.retry_exceptions:
+                self._attempts += 1
                 # Note: max_tries == -1 means infinite tries.
```
Comment (StephenSorriaux) [inline kazoo/retry.py:137]: Do you see a way to test this case?

File: kazoo/tests/test_retry.py (hunk 1 above)
Comment (ceache) [reply kazoo/tests/test_retry.py:6]: I cleaned this up.
I removed the class to get rid of the unittest model.
I think I cleaned this up enough now that it is actually better.

File: kazoo/retry.py (hunk 3 above)
Comment (ceache) [reply kazoo/retry.py:137]: I am not sure what you mean: what 'case'?
Do you mean that the exception is in the `self.exceptions` set?

Comment (StephenSorriaux) [reply kazoo/retry.py:137]: I was thinking of the not `Note: max_tries == -1 means infinite tries.`

**Review by StephenSorriaux — APPROVED** [review]

File: kazoo/tests/test_retry.py (hunk 1 above)
Comment (jeffwidman) [reply kazoo/tests/test_retry.py:6]: FWIW, I'm a fan of getting rid of the Unittest subclassing model whenever possibel... I personally find it clunky compared to the pytest syntax.

**Review by jeffwidman — APPROVED** [review]

**Pull request merged**